keps/sig-node/2727-grpc-probe/README.md

```diff
@@ -0,0 +1,167 @@
+# KEP-2727: Add GRPC Probe
+
+<!-- toc -->
+- [Release Signoff Checklist](#release-signoff-checklist)
+- [Goals](#goals)
+- [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+    - [Alpha](#alpha)
+    - [Beta](#beta)
+    - [GA](#ga)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+- [Alternatives](#alternatives)
+- [References](#references)
+<!-- /toc -->
+
+
+## Release Signoff Checklist
+
+- [ ] Enhancement issue in release milestone, which links to KEP dir in
+  [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] KEP approvers have approved the KEP status as `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture
+      and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in
+  [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents,
+  links to mailing list discussions/SIG meetings, relevant PRs/issues,
+  release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Goals
+
+Enable gRPC probe natively from Kubelet without requiring users to package a
+gRPC healthcheck binary with their container.
+
+- https://github.com/grpc-ecosystem/grpc-health-probe
+- https://github.com/grpc/grpc/blob/master/doc/health-checking.md
+
+## Non-Goals
+
+Add gRPC support in other areas of K8s (e.g. Services).
+
+## Proposal
+
+Host: pod ip
+
+```yaml
+    readinessProbe:
+      grpc:                     #+
+        port: 9090              #+
+        service: my-service     #+
```

Comment:
Don't most grpc services run with TLS? In fact IIRC it's part of the spec. Are we expecting services to expose their grpc health checks via HTTP rather than HTTPS with this?

Reply:
Add comment about using HTTP/2 as the default

Reply:
That doesn't really answer the question. Browsers require TLS for HTTP/2 but you can bypass it in grpc clients. Which are we doing? If we are requiring TLS, what client cert do we use? Do we validate server cert? What if the server validates client cert? We should make some choices and document them. Maybe if there is already ecosystem stuff for this we can borrow those decisions.

Reply:
I think you refer to these settings: https://github.com/grpc-ecosystem/grpc-health-probe/#health-checking-tls-servers

Reply:
search on GitHub uncovers a few usas for these parameters: https://github.com/search?q=grpc_health_probe+extension%3Ayaml+tls

Reply:
I assumed it is part of "wait for Beta feedback" among other customizations we may need to add

Reply:
Ok maybe note that here.

Reply:
Ah -- right I forgot the HTTP/2 can be non-TLS-- thinking back to the SPDY days.

Reply:
Update discussion to be more clear.

Reply:
thanks